### DISCUSSION TITLE
Changing how a mapped SQLAlchemy datetime field is serialized

### DISCUSSION THREAD

**Discussion by Keating950** [litestar-org/litestar#3783] [Q&A]
Hello all,

Let's say I have a SQLAlchemy ORM class that contains a datetime field. If I want to customize how this class is serialized (for example, to customize format), how would I go about this? Is this something handled at the DTO layer, or somewhere else?

```python
import datetime
from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column


class Base(DeclarativeBase): ...


class Foobar(Base):
    created_on: Mapped[datetime.datetime]
```

**Comment by vinicq** [comment]
Two places to hook in depending on how broad you want the change.

Project-wide, every datetime field across every response: `type_encoders` on the Litestar app.

```python
from datetime import datetime, timezone
from litestar import Litestar

def encode_dt(value: datetime) -> str:
    if value.tzinfo is None:
        value = value.replace(tzinfo=timezone.utc)
    return value.strftime("%Y-%m-%dT%H:%M:%SZ")

app = Litestar(
    route_handlers=[...],
    type_encoders={datetime: encode_dt},
)
```

Per-route or per-DTO. Define a DTO and override the encoding there. With `SQLAlchemyDTO` from `advanced-alchemy`:

```python
from advanced_alchemy.extensions.litestar import SQLAlchemyDTO, SQLAlchemyDTOConfig

class FoobarDTO(SQLAlchemyDTO[Foobar]):
    config = SQLAlchemyDTOConfig(rename_fields={"created_on": "createdOn"})
```

When the format varies per response (UTC vs local, microsecond precision toggle, ISO vs RFC1123), a route-scoped DTO with a custom serializer is cleaner than global encoders. You keep the freedom to vary per endpoint.

QA-side note that saved me once. If you change serialization format, write a snapshot test against the JSON bytes themselves. Easy to ship a local-time string by accident on a CI worker that defaults to UTC while developing on a machine that doesn't.
